#455: Integrate SPARQL Functions

simonstey opened the pull request:
Closes #455

* [See this document rendered online here](https://raw.githack.com/w3c/data-shapes/issue-455/shacl12-node-expr/index.html)

simonstey committed:
feat: :sparkles: automatically fetch sparql functions and create sections dynamically

simonstey commented:
Do we want to have the SPARQL function sections being populated dynamically (from https://github.com/w3c/sparql-query/blob/main/sparql-ns.ttl ) or just updated on demand?

HolgerKnublauch commented:
I am afraid the proper process would be:

1) We create a new TTL file that owl:imports the base file
2) In that file we add triples that are useful for rendering, e.g.:
    a) grouping by category
    b) deep links into the SPARQL spec
3) Then the content can be created dynamically, if we have a TTL parser available.
4) Maybe the TTL will be accepted by the sparql main document too, but I am not sure about the update cycle.

This is of course a lot of work but will be paying off on the long term because the info will be machine readable, e.g. for syntax-directed editors.

BTW the diff is huge - what happened?

simonstey commented:
> I am afraid the proper process would be:
> 
>     1. We create a new TTL file that owl:imports the base file
> 
>     2. In that file we add triples that are useful for rendering, e.g.:
>        a) grouping by category
>        b) deep links into the SPARQL spec
> 
>     3. Then the content can be created dynamically, if we have a TTL parser available.
> 
>     4. Maybe the TTL will be accepted by the sparql main document too, but I am not sure about the update cycle.
> 
> 
> This is of course a lot of work but will be paying off on the long term because the info will be machine readable, e.g. for syntax-directed editors.
> 
> BTW the diff is huge - what happened?

I messed up the formatting.. will fix that next

simonstey committed:
style: :art: restore formatting of sections

afs commented:
> 2. deep links into the SPARQL spec

If contributed back to RDF & SPARQL WG, I'm sure that could get incorporated into sparql-ns.

simonstey committed:
Import SPARQL function triples into shnex-sparql.ttl — labels, descriptions, categories, spec links.

simonstey committed:
Render SPARQL functions table in node-expr UI from local shnex-sparql.ttl

simonstey committed:
feat: Add examples for SPARQL functions in shnex-sparql vocabulary

- Introduced examples for various SPARQL functions to enhance documentation and usability.
- Added a new prefix for shnex-sparql to categorize examples.
- Updated existing functions with example usage, including arithmetic, logical, string, numeric, and date functions.
- Ensured all aggregate functions also include example usage for clarity.

HolgerKnublauch commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,901 @@
+	rdfs:label "W3C SHACL SPARQL Functions Vocabulary"@en ;
+	rdfs:comment "This vocabulary defines SPARQL functions for use with SHACL."@en ;
+	sh:declare [
+		sh:prefix "shnex" ;
```
Should declare sparql instead

simonstey replied:
addressed in https://github.com/w3c/data-shapes/pull/511/commits/530b33dc38d504091a4ed2bc4bd521855418f680

HolgerKnublauch commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,901 @@
+
+sparql:plus
+	a sparql:Function ;
+	rdfs:label "Plus"@en ;
```
I would use lower case for all these labels.

simonstey replied:
addressed in https://github.com/w3c/data-shapes/pull/511/commits/530b33dc38d504091a4ed2bc4bd521855418f680

HolgerKnublauch commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,901 @@
+sparql:plus
+	a sparql:Function ;
+	rdfs:label "Plus"@en ;
+	dc:description "This operator adds two numeric expressions and returns their sum."@en ;
```
I would use rdfs:comment - no need to introduce a new namespace just for that.

simonstey replied:
addressed in https://github.com/w3c/data-shapes/pull/511/commits/753311e3dec073ec49740accf58ea07e3087f25c

HolgerKnublauch commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,901 @@
+# Category values: algebraic, numeric, string, date, aggregate, other
+
+sparql:plus
+	a sparql:Function ;
```
Pending our final design, I believe this would become sh:ListParameterExpressionFunction.

simonstey replied:
this I took directly from https://github.com/w3c/sparql-query/blob/main/sparql-ns.ttl. Should I make them all `a sh:ListParameterExpressionFunction` instead?

HolgerKnublauch replied:
I believe yes, but we may change the final design later. We could also leave it like this and make sparql:Function a subclass of sh:ListParameterExpressionFunction. So for now, maybe just leave it.

HolgerKnublauch commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,901 @@
+	dc:description "This operator adds two numeric expressions and returns their sum."@en ;
+	shnex:category "algebraic" ;
+	rdfs:isDefinedBy <http://www.w3.org/TR/sparql12-query/#func-plus> ;
+	shnex-sparql:example """[ sparql:plus ( 38 4 ) ]""" ;
```
Another nice property here would be infixOperator "+" that can be used to convert into SPARQL syntax.

simonstey replied:
added infix property in https://github.com/w3c/data-shapes/pull/511/commits/fdfad059690e0b0d31b8a4deb10b8d12092dbc62 and added it to selected functions according to https://www.w3.org/TR/sparql12-query/#OperatorMapping

HolgerKnublauch replied:
Excellent job, mate!

simonstey committed:
chore: Normalize SPARQL function labels to lowercase in shnex-sparql.ttl; fix isdefinedby URIs

simonstey committed:
refactor: :speech_balloon: Replace dc:description with rdfs:comment

simonstey committed:
feat: add property definitions for function categorization, example and infix operator symbols. Add infix operators to selected functions

afs commented:
SPARQL functions work on arguments that are single RDF terms.  In `[ sparql:agg-count ( ?values ) ]` there isn't a term for `?values`. (The pain of lists not being first class in RDF.)

SPARQL Aggregates tied into the way `GROUP BY` works (e.g. `SUM(?X+1)`) to produce sets - the design means the lists are solely within aggregates a don't leak out to general evaluation. SPARQL Aggregates are not directly usable as node expressions.

shnex has it's own aggregates and that's probably the way to go - I think that defining an adaption layer to get SPARQL aggregates to work is not that different from shnex's aggregates so there is no advantage to SPARQL aggregates.

shnex aggregates working on lists which integrates with the way shnex is e.g. `sh:values`.

`[ sparql:coalesce ( ?optionalValue "default" ) ]` is n-ary including a "no result" error.
It is also special because of it's error handling - ditto `logical-and`, and `logical-or`.

I don't see that node expressions can have variables other than the environment : https://github.com/w3c/data-shapes/issues/498#issuecomment-3193749433

afs commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+shnex-sparql:infixOperator
+	a rdf:Property ;
+	rdfs:label "infix operator"@en ;
+	rdfs:comment "Indicates the infix operator symbol used in SPARQL syntax for this function, enabling conversion from functional form to infix notation."@en ;
```
Where is infix used in node expressions?

HolgerKnublauch replied:
It is not but the information is useful to capture IMHO, to produce human-readable syntax.

afs replied:
That would also need infix precedence.

simonstey replied:
> That would also need infix precedence.

which means? 🤔

robert-david replied:
@afs @simonstey i think this is still open?

HolgerKnublauch replied:
The infix predicate is optional and technical and shouldn't block this PR. We can turn it into a fresh issue if we want to pursue it forward.

afs commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+	shnex-sparql:example """[ sparql:plus ( 38 4 ) ]""" ;
+	.
+
+sparql:subtract
```
This has the same URI as the entry in `/ns/sparql.ttl`.

But both graphs have triples with subject `sparql:subtract` and properties `rdfs:comment` and `rdfs:isDefinedBy` with different object.

It would be better to only include the SHACL Node Expression additions.

HolgerKnublauch replied:
Yes, if necessary create a fresh property to avoid clashes.

simonstey replied:
@afs 
> This has the same URI as the entry in /ns/sparql.ttl.

what has the same URI? `shnex-sparql:example` ?

HolgerKnublauch replied:
I guess Andy's comment was about the subjects such as sparql:subtract. Both documents use rdfs:comment and rdfs:isDefinedBy. But IMHO those values should align. So if we in the SHACL WG find incorrect comments, the same fix should be applied to the /ns/sparql.ttl

Having said this, I don't think we need to owl:import the /ns/sparql.ttl graph so it will unlikely affect anyone.

In the interest of moving forward, I'll mark this as resolved but please reopen if anyone disagrees.

TallTed commented on shacl12-node-expr/index.html:
```diff
@@ -1,415 +1,747 @@
+				// Table
+				const table = document.createElement('table');
+				table.id = 'sparql-functions-table';
+				table.setAttribute('summary','List of SPARQL functions with description, spec link and category');
```
```suggestion
				table.setAttribute('summary','List of SPARQL functions with description, spec link, and category');
```

simonstey replied:
addressed in https://github.com/w3c/data-shapes/pull/511/commits/5479885720df5e81608f9016dfd64a98250a6110

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+sparql:filter-exists
+	a sparql:FunctionalForm ;
+	rdfs:label "filter exists"@en ;
+	rdfs:comment "This form tests whether a given pattern exists for each solution."@en ;
```
```suggestion
	rdfs:comment "This form tests whether a given pattern exists in each solution."@en ;
```

simonstey replied:
addressed in  https://github.com/w3c/data-shapes/pull/511/commits/5479885720df5e81608f9016dfd64a98250a6110

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+sparql:filter-not-exists
+	a sparql:FunctionalForm ;
+	rdfs:label "filter not exists"@en ;
+	rdfs:comment "This form tests whether a given pattern does not exist for each solution."@en ;
```
```suggestion
	rdfs:comment "This form tests that a given pattern does not exist in each solution."@en ;
```

simonstey replied:
addressed in  https://github.com/w3c/data-shapes/pull/511/commits/5479885720df5e81608f9016dfd64a98250a6110

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+	rdfs:comment "This function tests whether a string matches a regular expression pattern, optionally with a specified flag (e.g., i for case-insensitive)."@en ;
+	shnex-sparql:category "string" ;
+	rdfs:isDefinedBy <http://www.w3.org/TR/sparql12-query/#func-regex> ;
+	shnex-sparql:example """[ sparql:regex ( "hello123" "[0-9]+" ) ]""" ;
```
Where does the "specified flag" go? It seems this should have either multiple examples, or one that includes the optional "specified flag".

simonstey replied:
```suggestion
	shnex-sparql:example """[ sparql:regex ( "HeLLo123" "^hello" "i" ) ]""" ;
```

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+sparql:day
+	a sparql:Function ;
+	rdfs:label "day"@en ;
+	rdfs:comment "This function returns the day component of an xsd:dateTime or xsd:date."@en ;
```
```suggestion
	rdfs:comment "This function returns the day (1–31) component of an xsd:dateTime or xsd:date."@en ;
```

simonstey replied:
```suggestion
	rdfs:comment "This function returns the day (1-31) component of an xsd:dateTime or xsd:date."@en ;
```
shouldn't it be 1-31 then?

HolgerKnublauch replied:
Note: these comments were copied from the original source at https://github.com/w3c/sparql-query/blob/main/sparql-ns.ttl and should be addressed there.

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+sparql:month
+	a sparql:Function ;
+	rdfs:label "month"@en ;
+	rdfs:comment "This function returns the month component of an xsd:dateTime or xsd:date."@en ;
```
```suggestion
	rdfs:comment "This function returns the month component (1–12) of an xsd:dateTime or xsd:date."@en ;
```

simonstey replied:
```suggestion
	rdfs:comment "This function returns the month component (1-12) of an xsd:dateTime or xsd:date."@en ;
```

shouldn't it be 1-12 then?

HolgerKnublauch replied:
Note: these comments were copied from the original source at https://github.com/w3c/sparql-query/blob/main/sparql-ns.ttl and should be addressed there.

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,957 @@
+sparql:seconds
+	a sparql:Function ;
+	rdfs:label "seconds"@en ;
+	rdfs:comment "This function returns the second component (0–60, including leap seconds) of an xsd:dateTime value."@en ;
```
If this is "including leap seconds", the range must reach 61 (possibly higher; I do not know the maximum number of leap seconds that might be applied in any given year).
```suggestion
	rdfs:comment "This function returns the seconds component (0 ≤ seconds < 60, including leap seconds) of an xsd:dateTime value."@en ;
```

afs replied:
This counting from zero.
`60` would be the leap  second.

But `xsd:dateTime` does not support leap seconds.
"a decimal number greater than or equal to 0 and less than 60."
Discussion: https://www.w3.org/TR/xmlschema11-2/#d-t-values

simonstey replied:
in which case it should actually be:

```suggestion
	rdfs:comment "This function returns the second (0–59) component of an xsd:dateTime value."@en ;
```

afs replied:
`59.5` is legal.

simonstey replied:
so `[0,60)` then? 

what about removing the value range altogether?
```suggestion
	rdfs:comment "This function returns the seconds component of an xsd:dateTime value."@en ;
```

afs replied:
> what about removing the value range altogether?

+1

TallTed replied:
Updated my suggestion with `0 ≤ seconds < 60`

afs replied:
Remove the range. The function returns the seconds component of a valid xsd:dateTime(Stamp).

TallTed replied:
```suggestion
	rdfs:comment "This function returns the seconds component of a valid xsd:dateTime or xsd:dateTimeStamp value."@en ;
```

afs commented:
s/second/seconds/ to reduce the ambiguity of "the second component of".

simonstey committed:
fix: update comments for clarity in SPARQL functions and update .gitignore entries

HolgerKnublauch commented:
@simonstey this will be a great contribution, but the PR has been in draft state and unchanged for several weeks. Do you need input/help to finish it?

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -247,7 +247,7 @@ sparql:filter-exists
 	a sparql:FunctionalForm ;
 	rdfs:label "filter not exists"@en ;
-	rdfs:comment "This form tests whether a given pattern does not exist for each solution."@en ;
+	rdfs:comment "This form tests that a given pattern does not exist in each solution."@en ;
```
Upon re-reading, I'm not sure what is meant by this.

Is it testing that each (i.e., _every_) solution does not contain the given pattern?

Or that _some_ solution(s) do not contain the given pattern?

Or something else?

Probably, some rephrasing such as I've done here should be applied.

HolgerKnublauch replied:
This functional form does not fit into the node expression execution model, so it was removed alongside filter-exists, in and not-in. Please reopen a fresh issue if anyone disagrees.

simonstey commented:
@HolgerKnublauch 

> @simonstey this will be a great contribution, but the PR has been in draft state and unchanged for several weeks. Do you need input/help to finish it?

whoops, sry.. I'll have this PR merge-ready by next Monday!

simonstey committed:
fix: :twisted_rightwards_arrows: prepare rebase with HEAD

simonstey committed:
Merge branch 'gh-pages' into issue-455

simonstey committed:
fix: clean up HTML structure; fix merge issues

simonstey commented:
@HolgerKnublauch 
I rebased the document and fixed various HTML/merge commit issues. Should be good to go now. fyi, currently I only render the SPARQL functions as a table rather than fully fledged sections.

HolgerKnublauch committed:
Merge branch 'gh-pages' into issue-455

HolgerKnublauch committed:
Minor cleanup to align indentation, and make sure error message doesn't wipe out complete section

HolgerKnublauch committed:
Merged Description and Example into one column to (hopefully) improve layout

HolgerKnublauch committed:
Corrections to some examples

HolgerKnublauch commented:
Thanks a lot, Simon! I have made mostly cosmetic updates to your branch, which are hopefully OK.

We need to decide what to do with aggregates though. They are very different from the other functions as the don't work as in SPARQL: In a SPARQL query you would use them with a variable while in Node Expressions, the processing would work best if they receive another node expression that produces the aggregate values.

This however does not match how the other functions here work, which receive only individual values as arguments, never a list of values. Also, the shnex namespace already contains aggregates.

I suggest we delete the aggregates from our sparql: namespace.

TallTed commented on shacl12-vocabularies/shnex-sparql.ttl:
```diff
@@ -0,0 +1,958 @@
+sparql:greater-than-or-equal
+	a sparql:Function ;
+	rdfs:label "greater than or equal"@en ;
+	rdfs:comment "This operator tests whether the first RDF term is greater or equal to the second RDF term."@en ;
```
```suggestion
	rdfs:comment "This operator tests whether the first RDF term is greater than or equal to the second RDF term."@en ;
```

HolgerKnublauch committed:
Update shacl12-vocabularies/shnex-sparql.ttl

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

HolgerKnublauch committed:
Update shacl12-vocabularies/shnex-sparql.ttl

Co-authored-by: Ted Thibodeau Jr <tthibodeau@openlinksw.com>

HolgerKnublauch committed:
Replaced rdf:types of functions to follow SHACL conventions

The classes sparql:Function etc are ill-defined and serve IMHO no purpose from a SHACL point of view. By replacing them with sh:ListParameterExpressionFunction we follow SHACL syntax and also make the owl:import to the external namespace unnecessary. (That owl:import does not resolve right now anyway)

HolgerKnublauch committed:
Merge branch 'issue-455' of https://github.com/w3c/data-shapes into issue-455

HolgerKnublauch committed:
Deleted aggregate functions because they belong into shnex as they take lists as parameters

HolgerKnublauch committed:
Removed filter-exists, not-exists, in and not-in

These do not fit the execution model of node expressions. If people still want to use them, they can use SPARQL strings via sh:select etc.

HolgerKnublauch committed:
Tabs vs spaces

HolgerKnublauch committed:
Updated JS to new rdf:type

HolgerKnublauch reviewed (approved):
To me it's good enough for now. Further refinements can and will happen in fresh tickets, when needed.

HolgerKnublauch committed:
Merge branch 'gh-pages' into issue-455

simonstey commented:
@robert-david

robert-david committed:
Minor fixes

robert-david commented:
@simonstey should we specify what is returned for each of the sparql functions to streamline the descriptions? Currently, sometimes it is stated and sometimes not. For some functions, it might not be clear.
